src/util/sighash.rs

```diff
@@ -0,0 +1,904 @@
+// Rust Bitcoin Library
+// Written in 2021 by
+//   The rust-bitcoin developers
+//
+// To the extent possible under law, the author(s) have dedicated all
+// copyright and related and neighboring rights to this software to
+// the public domain worldwide. This software is distributed without
+// any warranty.
+//
+// You should have received a copy of the CC0 Public Domain Dedication
+// along with this software.
+// If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+//
+
+//! # Generalized, efficient, signature hash implementation
+//!
+//! Implementation of the algorithm to compute the message to be signed according to [Bip341](https://github.com/bitcoin/bips/blob/150ab6f5c3aca9da05fccc5b435e9667853407f4/bip-0341.mediawiki),
+//! [Bip143](https://github.com/bitcoin/bips/blob/99701f68a88ce33b2d0838eb84e115cef505b4c2/bip-0143.mediawiki)
+//! and legacy
+//!
+
+pub use blockdata::transaction::SigHashType as LegacySigHashType;
+use consensus::{encode, Encodable};
+use core::fmt;
+use core::ops::{Deref, DerefMut};
+use hashes::{sha256, sha256d, Hash};
+use io;
+use util::taproot::{TapLeafHash, TapSighashHash};
+use SigHash;
+use {Script, Transaction, TxOut};
+
+use prelude::*;
+
+/// Efficiently calculates signature hash message for legacy, segwit and taproot inputs.
+#[derive(Debug)]
+pub struct SigHashCache<T: Deref<Target = Transaction>> {
```

Comment:
Not sure I understand why this does not use `AsRef<Transaction>` instead of `Deref<Target = Transaction>`.

Reply:
Or `Borrow` which would also allow for `SigHashCache<Transaction>` - not sure how important it is.

Reply:
We can't use `AsRef` because we need also a mutable borrow for `access_witness` and `AsMut` is not a subtrait of `AsRef` (while we have `DerefMut: Deref`)

We could use `Borrow` and `BorrowMut` with the small drawback to call `borrow()` every time (or use a function returning `&Transaction`). The other downside is that we can't provide an impl for deprecated `bip143::SigHashCache` like we do now.

What are the downsides of using `Deref`? (Apart from being discouraged to use by guidelines to emulate object oriented, which doesn't seem to be the case here)

Reply:
> `AsMut` is not a subtrait of `AsRef`

I don't see why that would be a problem. Just bound it by `AsRef<Transaction> + AsMut<Transaction>` where needed.

> What are the downsides of using `Deref`?

It's less general (e.g. `Box<PartiallySignedTransaction>` could theoretically implement both `AsRef<PartiallySignedTransaction>` and `AsRef<Transaction>` but it can never implement both `Deref`s), but honestly, I don't know how relevant it is. It's also a bit of stylistic issue - code in `std` doesn't use `Deref` in bounds.

I personally don't feel strongly about it.

Reply:
> I don't see why that would be a problem. Just bound it by AsRef<Transaction> + AsMut<Transaction> where needed.

Maybe you want to use `SigHashCache` having only the non-mutable borrow of `Transaction` and you don't need to use `access_witness`

Reply:
I agree `AsRef` + `AsMut` would be slightly better but it would make it worse to implement the deprecated method in `bip143.rs` that is using `Deref` and `DerefMut`. So I am leaving like it is unless someone feels strongly about it.